File: src/com/tao/chinachuclient/MainActivity.java
package com.tao.chinachuclient;

import java.util.ArrayList;

import com.tao.chinachuclient.data.Server;

import Chinachu4j.Chinachu4j;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Base64;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;

public class MainActivity extends Activity implements OnItemClickListener{

	private SharedPreferences pref;
	private Chinachu4j chinachu;
	private ApplicationClass appClass;
	private String chinachuAddress, username, password;
	private ListView mainList;

	@Override
	protected void onCreate(Bundle savedInstanceState){
		super.onCreate(savedInstanceState);
		mainList = new ListView(this);
		setContentView(mainList);

		getActionBar().setDisplayShowHomeEnabled(false);

		pref = PreferenceManager.getDefaultSharedPreferences(this);
		appClass = (ApplicationClass)getApplicationContext();

		chinachuAddress = pref.getString("chinachuAddress", null);
		username = pref.getString("username", null);
		password = pref.getString("password", null);
		if(chinachuAddress == null || username == null || password == null) {
			Intent i = new Intent(this, AddServer.class);
			i.putExtra("startMain", true);
			startActivity(i);
			finish();
			return;
		}
		username = new String(Base64.decode(username, Base64.DEFAULT));
		password = new String(Base64.decode(password, Base64.DEFAULT));

		String[] listItem = new String[]{"番組表", "ルール", "予約済み", "録画中", "録画済み"};
		ArrayAdapter<String> adapter = new ArrayAdapter<String>(this, android.R.layout.simple_list_item_1, listItem);
		mainList.setAdapter(adapter);
		mainList.setOnItemClickListener(this);

		chinachu = new Chinachu4j(chinachuAddress, username, password);

		appClass.setChinachu(chinachu);

		appClass.setStreaming(pref.getBoolean("streaming", false));
		appClass.setEncStreaming(pref.getBoolean("encStreaming", false));
	}

	@Override
	public void onItemClick(AdapterView<?> parent, View view, int position, long id){
		if(position == 0)
			startActivity(new Intent(this, ChannelScheduleActivity.class));
		else if(position == 1)
			startActivity(new Intent(this, RuleActivity.class));
		else if(position > 1) {
			Intent i = new Intent(this, ProgramActivity.class);
			i.putExtra("type", position);
			startActivity(i);
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu){
		MenuItem changeServer = menu.add(0, Menu.FIRST, Menu.NONE, "鯖変更");
		changeServer.setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM);
		MenuItem item = menu.add(0, Menu.FIRST + 1, Menu.NONE, "設定");
		item.setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item){
		if(item.getItemId() == Menu.FIRST) {
			final DBUtils dbUtils = new DBUtils(this);
			final ArrayList<String> address = new ArrayList<String>();
			Server[] servers = dbUtils.getServers();
			for(Server s : servers)
				address.add(s.getChinachuAddress());

			int settingNow = address.indexOf(pref.getString("chinachuAddress", ""));
			new AlertDialog.Builder(this).setTitle("サーバーを選択してください")
			.setSingleChoiceItems((String[])address.toArray(new String[0]), settingNow, new OnClickListener(){
				@Override
				public void onClick(DialogInterface dialog, int which){
					String selectedAddress = address.get(which);
					Server server = dbUtils.getServerFromAddress(selectedAddress);
					dbUtils.serverPutPref(MainActivity.this, server);

					chinachu = new Chinachu4j(server.getChinachuAddress(),
							new String(Base64.decode(server.getUsername(), Base64.DEFAULT)),
							new String(Base64.decode(server.getPassword(), Base64.DEFAULT)));
					appClass.setChinachu(chinachu);
					appClass.setStreaming(server.getStreaming());
					appClass.setStreaming(server.getEncStreaming());

					dbUtils.close();
					dialog.dismiss();
				}
			}).setPositiveButton("キャンセル", null).show();
		}else if(item.getItemId() == Menu.FIRST + 1){
			startActivity(new Intent(this, Preference.class));
		}
		return super.onOptionsItemSelected(item);
	}
}

File: app/src/main/java/com/tao/chinachuclient/MainActivity.kt
package com.tao.chinachuclient

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.PreferenceManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class MainActivity : AppCompatActivity(), AdapterView.OnItemClickListener {

    private lateinit var app: App
    private lateinit var mainList: ListView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mainList = ListView(this)
        setContentView(mainList)

        val pref = PreferenceManager.getDefaultSharedPreferences(this)
        app = applicationContext as App

        CoroutineScope(Dispatchers.Main).launch {
            val chinachuAddress = pref.getString("chinachuAddress", "") ?: ""
            val serverExists = withContext(Dispatchers.IO) { app.serverRepository.isExists(chinachuAddress) }
            if (chinachuAddress == "" || !serverExists) {
                startActivity(Intent(this@MainActivity, AddServer::class.java).apply {
                    putExtra("startMain", true)
                })
                finish()
                return@launch
            }

            withContext(Dispatchers.IO) { app.serverRepository.findByAddress(chinachuAddress) }?.let {
                app.changeCurrentServer(it)
            }

            val listItem = resources.getStringArray(R.array.main_list_names)
            val adapter = ArrayAdapter(this@MainActivity, android.R.layout.simple_list_item_1, listItem)
            mainList.adapter = adapter
            mainList.onItemClickListener = this@MainActivity
        }
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        startActivity(when (position) {
            Type.CHANNEL_SCHEDULE_ACTIVITY -> Intent(this, ChannelScheduleActivity::class.java)
            Type.RULE_ACTIVITY -> Intent(this, RuleActivity::class.java)
            else -> Intent(this, ProgramActivity::class.java).putExtra("type", position)
        })
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menu?.add(0, Menu.FIRST, Menu.NONE, R.string.change_server)?.setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
        menu?.add(0, Menu.FIRST + 1, Menu.NONE, R.string.settings)?.setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == Menu.FIRST) {
            CoroutineScope(Dispatchers.Main).launch {
                val servers = withContext(Dispatchers.IO) { app.serverRepository.getAll() }
                val address = servers.map { it.chinachuAddress }

                val currentServer = app.currentServer
                val settingNow = address.indexOf(currentServer.chinachuAddress)
                AlertDialog.Builder(this@MainActivity)
                    .setTitle(R.string.select_server)
                    .setSingleChoiceItems(address.toTypedArray(), settingNow) { dialog, which ->
                        CoroutineScope(Dispatchers.Main).launch {
                            val selectedAddress = address[which]
                            withContext(Dispatchers.IO) { app.serverRepository.findByAddress(selectedAddress) }?.let {
                                app.changeCurrentServer(it)
                            }
                            dialog.dismiss()
                        }
                    }
                    .setPositiveButton(R.string.cancel, null)
                    .show()
            }
        } else if (item.itemId == Menu.FIRST + 1) {
            startActivity(Intent(this, Preference::class.java))
        }
        return super.onOptionsItemSelected(item)
    }
}
